### ISSUE TITLE
Implement full-node syncing

### ISSUE THREAD

**Issue by gsalgado** [ethereum/py-evm#330]
This is a high-level overview of full-node syncing, heavily inspired by the way geth does it. I'm going to split this into smaller tasks, and create separate issues for them. This issue is just to help keep track of overall progress

There will be two modes of operation (fast and normal), which can be selected via command line flags

#### Fast-sync:
* download headers and bodies
* verify header chain's consistency (PoW, etc)
* download transaction receipts
* when at head - 1024 blocks, pause chain sync and download the state trie defined by the root hash of current head
* switch to normal sync

#### Normal-sync:
* download headers and bodies
* verify header chain's consistency (PoW, etc)
* import blocks

### Starting the sync

Upon connecting to a new ETH node, trigger a sync using the peer with the highest
announced TD (total difficulty).

Upon receiving a NewBlock msg with a TD that's higher than the peer's previously announced TD,
trigger a sync using that peer.

The sync method will return immediately if there is a sync in progress or if we have less than
`min_peers_to_sync` connected peers

### The sync method

The sync runs in a loop, querying the given peer for `MAX_HEADERS_FETCH` headers starting from
the current local head, skipping 128 headers between each (aka a header skeleton). It then makes
concurrent requests, using all available peers, to fill the gaps in that skeleton. Once all gaps
in that skeleton are filled, the headers are processed, stored in the DB and the fetchers for block
bodies (and receipts, in case we're in fast mode) are notified that there are new headers.

The fetchers will then request the bodies/receipts for those new headers, in parallel, using all
available peers.

### Sanity checking after completing the fast sync

Since the fast-sync just downloads block/state data and stores it in the DB, it might be a good idea to iterate through all headers in the chain confirming that we have their transactions/uncles/receipts in the DB, as well as iterating over all state trie nodes starting from the root hash. Upon successful completion of that sanity check, we should be able to start normal syncing

### Further optimizations

These are optimizations that should be implemented in subsequent iterations

#### K-th header PoW verification

PoW verification is still a bottleneck even when in fast-sync mode, so instead of verifying the
PoW on every header, geth does so on a randomly chosen subset of them, in a way that the
probability of a forgery is smaller than the probability of an attacker producing a SHA3
collision. We'll need to look into doing that as well

#### Fine-grained peer idleness 

Keep separate idle flags (NodeIdle, ReceiptIdle, etc) for every peer, so that we only make,
say, a NodeData request to a peer if there's no pending NodeData requests for that peer.

**Comment by pipermerriam** [comment]
cc @lrettig
